p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,516 @@
+}
+
+var _ tpt.Transport = &WebRTCTransport{}
+
+type Option func(*WebRTCTransport) error
+
+type IceTimeouts struct {
+	Disconnect time.Duration
+	Failed     time.Duration
+	Keepalive  time.Duration
+}
+
+// WithPeerConnectionIceTimeouts sets the ice disconnect, failure and keepalive timeouts
+func WithPeerConnectionIceTimeouts(timeouts IceTimeouts) Option {
+	return func(t *WebRTCTransport) error {
+		if timeouts.Disconnect == 0 {
+			timeouts.Disconnect = t.peerConnectionTimeouts.Disconnect
+		}
+		if timeouts.Failed == 0 {
+			timeouts.Failed = t.peerConnectionTimeouts.Failed
+		}
+		if timeouts.Keepalive == 0 {
+			timeouts.Keepalive = t.peerConnectionTimeouts.Keepalive
+		}
+		// 0 is not treated as a default, and instead disables the respective check
+		// (eg. if keepalive interval is 0, no keepalives are sent, or if disconnected timeout is 0, the connection never enters a disconnected state).
+		// Refer here: https://github.com/pion/ice/blob/v2.3.0/agent_config.go#L73-L85
+		//
+		// The timeouts are only set to their default values if they are not set in the settingEngine.
+		// Refer here: https://github.com/pion/ice/blob/67f28cf23a8ae59a38e0128390ccdad27b5526be/agent_config.go#L213-L224
+		//
+		// Our implementation requires that the values be set (explicitly or implicitly)
+		// so that we can fail connections are free up resources if a remote disconnects abruptly,
+		// therefore 0 values are not acceptable. We also need to enforce that Failed timeout > disconnected timeout,
+		// and failed timeout > keepalive interval. (We currently enforce keepalive interval <= disconnected interval)
+		if timeouts.Disconnect != 0 {
+			if timeouts.Failed != 0 && timeouts.Failed < timeouts.Disconnect {
+				return fmt.Errorf("disconnect timeout cannot be greater than failed timeout")
+			}
+			if timeouts.Keepalive != 0 && timeouts.Disconnect <= timeouts.Keepalive {
+				return fmt.Errorf("keepalive timeout cannot be greater than or equal to failed timeout")
+			}
+		}
+		t.peerConnectionTimeouts = timeouts
+		return nil
+	}
+}
+
+// WithListenerMaxInFlightConnections sets the maximum number of connections that are in-flight, i.e
+// they are being negotiated, or are waiting to be accepted.
+func WithListenerMaxInFlightConnections(m uint32) Option {
+	return func(t *WebRTCTransport) error {
+		if m == 0 {
+			t.maxInFlightConnections = DefaultMaxInFlightConnections
+		} else {
+			t.maxInFlightConnections = m
+		}
+		return nil
+	}
+}
+
+func New(privKey ic.PrivKey, psk pnet.PSK, gater connmgr.ConnectionGater, rcmgr network.ResourceManager, opts ...Option) (*WebRTCTransport, error) {
+	if psk != nil {
+		log.Error("WebRTC doesn't support private networks yet.")
+		return nil, fmt.Errorf("WebRTC doesn't support private networks yet")
+	}
+	localPeerID, err := peer.IDFromPrivateKey(privKey)
+	if err != nil {
+		return nil, fmt.Errorf("get local peer ID: %w", err)
+	}
+	// We use elliptic P-256 since it is widely supported by browsers.
```

Comment:
What about ed25519?

Reply:
Will take a look as I do not immediately see it in my notes.

Reply:
This is a list of widely supported algorithms for `RTCPeerConnection`. https://w3c.github.io/webcrypto/#algorithm-recommendations-implementers . From my observation, RSA and ECDSA P-256 is supported on almost all browsers. Ed25519 is not present on the list.

Reply:
Thanks!